File: 0174-dungeon-game-地下城游戏/0174.py
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        col = len(dungeon)
        if not col:
            return
        row = len(dungeon[0])
        dp = [max(1, 1-dungeon[col - 1][row - 1])]
        for y in range (col-2, -1, -1):
            dp.insert(0, max(1, dp[0] - dungeon[y][row - 1]))
        for x in range (row-2, -1, -1):
            dp[col-1] = max(1, dp[col-1] - dungeon[col-1][x])
            for y in range (col-2, -1, -1):
                dp[y] = max(1, min(dp[y] - dungeon[y][x], dp[y+1] - dungeon[y][x]))
                
        return dp[0]
        

File: 0174-dungeon-game-地下城游戏/0174.cpp
class Solution {
    //void 
public:
    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int row = dungeon.size();
        int col = dungeon[0].size();
        if (row == 0 || col == 0) {
            return 1;
        }
        vector<vector<int>> hp = dungeon;
        for (int x = row - 1; x >= 0; --x) {
            for (int y = col - 1; y >=0; --y) {
                if (x == row - 1 && y == col - 1) {
                    hp[x][y] = max(1, 1 - dungeon[x][y]);
                } else if (x == row - 1) {
                    hp[x][y] = max(1, hp[x][y+1] - dungeon[x][y]);
                } else if (y == col - 1) {
                    hp[x][y] = max(1, hp[x+1][y] - dungeon[x][y]);
                } else {
                    int t = min(hp[x+1][y], hp[x][y+1]) - dungeon[x][y];
                    hp[x][y] = max(1, t);
                }
            }
        }
        return hp[0][0];
    }
};